core/services/pipeline/runner.go

```diff
@@ -244,68 +222,76 @@ func (m *memoryTaskRun) results() (a []Result) {
 	return
 }
 
-func (r *runner) ExecuteRun(ctx context.Context, run Run, l logger.Logger) (trrs TaskRunResults, err error) {
-	return r.executeRun(ctx, r.orm.DB(), run, l)
+func (r *runner) ExecuteRun(ctx context.Context, spec Spec, l logger.Logger) (trrs TaskRunResults, err error) {
+	return r.executeRun(ctx, r.orm.DB(), spec, l)
 }
 
-func (r *runner) executeRun(ctx context.Context, txdb *gorm.DB, run Run, l logger.Logger) (trrs TaskRunResults, err error) {
-	l.Debugw("Initiating tasks for pipeline run", "runID", run.ID)
+func (r *runner) executeRun(ctx context.Context, txdb *gorm.DB, spec Spec, l logger.Logger) (TaskRunResults, error) {
+	l.Debugw("Initiating tasks for pipeline run of spec", "spec", spec.ID)
+	var (
+		err  error
+		trrs TaskRunResults
+	)
 	startRun := time.Now()
 
-	if run.PipelineSpec.ID == 0 {
-		return nil, errors.Errorf("run.PipelineSpec.ID may not be 0: %#v", run)
+	d := TaskDAG{}
+	err = d.UnmarshalText([]byte(spec.DotDagSource))
+	if err != nil {
+		return trrs, err
 	}
 
+	// HACK: This mutex is necessary to work around a bug in the pq driver that
+	// causes concurrent database calls inside the same transaction to fail
+	// with a mysterious `pq: unexpected Parse response 'C'` error
+	// FIXME: Get rid of this by replacing pq with pgx
+	var txdbMutex sync.Mutex
```

Comment:
Looking at this now it might be worth creating an encapsulating struct that includes both the db and the mutex, what do you think?

Reply:
yep